### DISCUSSION TITLE
Godot's Dictionary  and C# builtin Dictionary Hash behaviors should be consistent

### DISCUSSION THREAD

**Discussion by KarryCharon** [godotengine/godot-proposals#9593] [Scripting]
### Describe the project you are working on

-

### Describe the problem or limitation you are having in your project

C# builtin Dictionary and Godot's Dictionary hash behaviors are not consistent.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

```c#
// GameplayTag has override GetHashCode and equal == != operators!
var tag1 = new GameplayTag("Player.Skill.Attack");
var tag2 = new GameplayTag("Player.Skill.Attack");
bool IsEqual1 = tag1 == tag2; // True
bool IsEqual2 = tag1.Equals(tag2); // True
bool IsHashCodeEqual = tag1.GetHashCode() == tag2.GetHashCode(); // True
var d1 = new System.Collections.Generic.Dictionary<GameplayTag, int>();
d1.Add(tag1, 1);
var d2 = new Godot.Collections.Dictionary<GameplayTag, int>();
d2.Add(tag1, 1);
bool IsInD1 = d1.ContainsKey(tag2); // True
bool IsInD2 = d2.ContainsKey(tag2); // False (Should be True)
```

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

-

### If this enhancement will not be used often, can it be worked around with a few lines of script?

-

### Is there a reason why this should be core and not an add-on in the asset library?

-

**Comment by raulsntos** [comment]
Godot collections don't use any of the C# comparison/equality methods (`GetHashCode`, `Equals`, `==` or `!=`).

When calling `Godot.Collections.Dictionary.ContainsKey` it just calls the native C++ implementation of the `Dictionary::has` method which uses the Variant hasher. For `GodotObject` and derived types this means calculating a hash from the object's pointer, so naturally different instances will have a different hash.

I'm not aware of any way to override this, so you'll have to convert the Godot collection to a C# collection if you want to use the C# comparison/equality.

**Comment by KarryCharon** [comment]
> Godot collections don't use any of the C# comparison/equality methods (`GetHashCode`, `Equals`, `==` or `!=`).
> 
> When calling `Godot.Collections.Dictionary.ContainsKey` it just calls the native C++ implementation of the `Dictionary::has` method which uses the Variant hasher. For `GodotObject` and derived types this means calculating a hash from the object's pointer, so naturally different instances will have a different hash.
> 
> I'm not aware of any way to override this, so you'll have to convert the Godot collection to a C# collection if you want to use the C# comparison/equality.

C# builtin Dictionary cant be export to Inspector, Godot's Dictionary cant rewrite hash for ContainsKey.
Both of them are important for me. :(

**Reply by raulsntos** [reply]
You can use both Dictionary types and convert between them as needed.

```csharp
// Convert System Dictionary to Godot Dictionary.
var systemDictionary = new System.Collections.Generic.Dictionary<GameplayTag, int>();
var godotDictionary = new Godot.Collections.Dictionary<GameplayTag, int>(systemDictionary);

// Convert Godot Dictionary to System Dictionary.
var godotDictionary = new Godot.Collections.Dictionary<GameplayTag, int>();
var systemDictionary = new System.Collections.Generic.Dictionary<GameplayTag, int>(godotDictionary);
```

**Reply by KarryCharon** [reply]
Yes it works. But there will still be additional performance cost. Looking forward to more elegant implementations in the future

**Reply by KarryCharon** [reply]
To expose GodotObject hash method to c# side may be more easier?
